lib/admin/parsers/practice.rb

```diff
@@ -1,49 +1,31 @@
 module Admin
   module Parsers
+
+    # Parse a string of practices into an array of hashes.
     class Practice
       RECORD_SEPARATOR = "\n"
-      PART_SEPARATOR = '|'
-      EMAIL_REGEX = /@/i
-      TEL_REGEX = /[\d\s]{8,12}/
-      POSTCODE_REGEX = /((GIR\s*0AA)|((([A-PR-UWYZ][0-9]{1,2})|(([A-PR-UWYZ][A-HK-Y][0-9]{1,2})|(([A-PR-UWYZ][0-9][A-HJKSTUW])|([A-PR-UWYZ][A-HK-Y][0-9][ABEHMNPRVWXY]))))\s*[A-Z]?[0-9][ABD-HJLNP-UW-Z]{2}))/i
-      URL_REGEX = /\A((http|https):\/\/)?[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,5}(([0-9]{1,5})?\/.*)?\Z/i
-
-      class << self
-        def parse(practices_data)
-          split_into_lines(practices_data).map do |practice_line|
-            identify_parts(practice_line)
-          end
-        end
-
-        def split_into_lines(practices_data)
-          practices_data && practices_data.split(RECORD_SEPARATOR) || []
-        end
 
-        def identify_parts(practice_line)
-          parts = split_into_parts(practice_line)
-          parts = parts.map { |part| sanitise(part) }
-          build_practice_hash(parts)
-        end
+      attr_reader :warnings
 
-        def build_practice_hash(parts)
+      def initialize
+        @warnings = []
+      end
 
-          parts.inject({}) do |result, part|
-            result[:email] = part if part.match(EMAIL_REGEX)
-            result[:address] = part if part.match(POSTCODE_REGEX)
-            result[:tel] = part if part.match(TEL_REGEX)
-            result[:url] = part if part.match(URL_REGEX)
-            result
-          end
+      def parse(practices_data, line_parser = PracticeLine.new)
+        practices = split_into_lines(practices_data).map.each_with_index do |practice_line, index|
+          line_parser.parse(practice_line, index+1)
```

Comment:
The `PracticeLine` class is confusingly named. I expected that each instance of `PracticeLine` would act upon a single practice line. It appears on closer inspection that `PracticeLine` isn't instantiated with a line of practice data. Instead, one class is instantiated and the `parse` method is called multiple times with `practice_line` and an index, which is purely there should there be a warning to output.

I think either the name needs to change to communicate better what's going on or the code needs to change, e.g. something like this:

```ruby
def parse(practices_data)
  practices = split_into_lines(practices_data)
  practices.map.each do |practice|
    practice_line = PracticeLine.new(practice)
    practice_line.parse
  end
end
```

The responsibility of generating the warnings could lie with this class. All it needs it access to the unmatched parts from each practice_line instance.